server/src/core/user/user-service.ts

```diff
@@ -138,10 +169,54 @@ export class Users implements UserService {
     }
   }
 
+  async nominateMiddleman(args: UserNominateMiddlemanArgs): Promise<User> {
+    validators.validatesNominateMiddleman(args);
+    const { phone, nominator } = args;
+    try {
+      const nominatorUser = await this.collection.findOne({ _id: nominator, roles: 'donor' });
+      if (!nominatorUser) throw createMiddlemanNominationFailedError();
+
+      const result = await this.collection.findOneAndUpdate(
+        { phone },
+        {
+          $addToSet: { roles: 'middleman', middlemanFor: nominator },
+          $currentDate: { updatedAt: true },
+          $setOnInsert: {
+            _id: generateId(),
+            password: '',
+            phone,
+            addedBy: nominator,
+            createdAt: new Date()
+          }
+        },
+        { upsert: true, returnOriginal: false, projection: NOMINATED_USER_PROJECTION }
+      );
+      return getSafeUser(result.value);
+    }
+    catch (e) {
+      if (e instanceof AppError) throw e;
+      if (e.code == 11000 && e.message.indexOf(args.phone) >= 0) {
```

Comment:
In this case, that's not an `AppError`, that's MongoDB error, and that's a MongoDB-specific error code. But you're right, we should create a descriptive constant for it.

Reply:
I've created a `MONGO_ERROR_DUPLICATE_KEY` constant for that error code and I've also created a `isMongoDuplicateKeyError()` function to check for duplicate-key errors instead of doing this manual check everywhere.